graph/api/src/main/scala/org/apache/spark/graph/api/CypherSession.scala

```diff
@@ -0,0 +1,285 @@
+   * @since 3.0.0
+   */
+  def extractRelationshipDatasets(relationships: Dataset[Row]): Set[RelationshipDataset] = {
+    val relColumns = relationships.columns.toSet
+    val relTypeColumns = relColumns.filter(_.startsWith(LABEL_COLUMN_PREFIX))
+    validateLabelColumns(relationships.schema, relTypeColumns)
+    val idColumns = Set(ID_COLUMN, SOURCE_ID_COLUMN, TARGET_ID_COLUMN)
+    val propertyColumns = relColumns -- idColumns -- relTypeColumns
+    val relProperties = propertyColumns.map(col => col -> col).toMap
+    relTypeColumns.map { relTypeColumn =>
+      val predicate = relationships.col(relTypeColumn)
+
+      RelationshipDataset(
+        relationships.filter(predicate),
+        ID_COLUMN,
+        SOURCE_ID_COLUMN,
+        TARGET_ID_COLUMN,
+        relTypeColumn.substring(1),
+        relProperties)
+    }
+  }
+
+  private def validateLabelColumns(schema: StructType, columns: Set[String]): Unit = {
+    schema.fields.filter(f => columns.contains(f.name)).foreach(field => {
+      if (field.dataType != BooleanType) {
+        throw new IllegalArgumentException(s"Column ${field.name} must be of type BooleanType.")
+      }
+    })
+  }
+
+}
+
+/**
+ * A CypherSession allows for creating, storing and loading [[PropertyGraph]] instances as well as
+ * executing Cypher queries on them.
+ *
+ * Wraps a [[org.apache.spark.sql.SparkSession]].
+ *
+ * @since 3.0.0
+ */
+@Evolving
+trait CypherSession {
+
+  def sparkSession: SparkSession
+
+  /**
+   * Executes a Cypher query on the given input graph.
+   *
+   * @param graph [[PropertyGraph]] on which the query is executed
+   * @param query Cypher query to execute
+   * @since 3.0.0
+   */
+  def cypher(graph: PropertyGraph, query: String): CypherResult
+
+  /**
+   * Executes a Cypher query on the given input graph.
+   *
+   * Note that queries can take optional parameters:
+   *
+   * {{{
+   *     Parameters:
+   *
+   *     {
+   *        "name" : "Alice"
+   *     }
+   *
+   *     Query:
+   *
+   *     MATCH (n:Person)
+   *     WHERE n.name = $name
+   *     RETURN n
+   * }}}
+   *
+   * @param graph      [[PropertyGraph]] on which the query is executed
+   * @param query      Cypher query to execute
+   * @param parameters parameters used by the Cypher query
+   * @since 3.0.0
+   */
+  def cypher(graph: PropertyGraph, query: String, parameters: Map[String, Any]): CypherResult
+
+  /**
+   * Executes a Cypher query on the given input graph.
+   *
+   * Note that queries can take optional parameters:
+   *
+   * {{{
+   *     Parameters:
+   *
+   *     {
+   *        "name" : "Alice"
+   *     }
+   *
+   *     Query:
+   *
+   *     MATCH (n:Person)
+   *     WHERE n.name = $name
+   *     RETURN n
+   * }}}
+   *
+   * @param graph      [[PropertyGraph]] on which the query is executed
+   * @param query      Cypher query to execute
+   * @param parameters parameters used by the Cypher query
+   * @since 3.0.0
+   */
+  def cypher(
```

Comment:
I don't want to cause further delay of this PR. But we should consider, maybe in a follow-up PR. It is more natural for users to start from a graph: 

~~~scala
graph.cypher(queryString).param(...).query()
~~~

Then the method here is mostly for the backend to implement cypher query. If that is the case, do we need overloading methods and Java compatibility? I think it is okay to say only Scala is supported in implementing the Cypher backend.

Reply:
Nice idea. It might simplify the API when we just delegate to the `CypherSession` within the query method. We can sketch it in the implementation PR and see how well it works, wdyt?